File: apps/api/coinmark_api/api/routes/bot.py
from __future__ import annotations

from fastapi import APIRouter, Query

from coinmark_api.services.bot.funding import get_funding_rate_top
from coinmark_api.services.bot.long_short_volume import get_long_short_volume_rank
from coinmark_api.services.bot.oi_marketcap import get_oi_marketcap_rank


router = APIRouter()


@router.get("/fundingRateTop")
async def funding_rate_top(
    limit: int = Query(15, ge=1, le=200),
    order: str = Query("abs", pattern="^(abs|desc|asc)$"),
) -> dict:
    data = await get_funding_rate_top(limit=limit, order=order)
    return {"items": data}


@router.get("/longShortVolumeRank")
async def long_short_volume_rank(
    market: str = Query("swap", pattern="^(spot|swap)$"),
    bucket: str = Query("15m", pattern="^(15m|1h|1d)$"),
    limit: int = Query(10, ge=1, le=200),
) -> dict:
    data = await get_long_short_volume_rank(market=market, bucket=bucket, limit=limit)
    return data


@router.get("/oiMarketCapRank")
async def oi_marketcap_rank(
    limit: int = Query(15, ge=1, le=200),
) -> dict:
    data = await get_oi_marketcap_rank(limit=limit)
    return {"items": data}

File: apps/api-go/internal/handler/bot.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"coinmark/api-go/internal/service"
)

func registerBotRoutes(g *gin.RouterGroup, d *Deps) {
	g.GET("/fundingRateTop", handleFundingRateTop(d))
	g.GET("/longShortVolumeRank", handleLSVolumeRank(d))
	g.GET("/oiMarketCapRank", handleOIMcapRank(d))
}

func handleFundingRateTop(d *Deps) gin.HandlerFunc {
	return func(c *gin.Context) {
		limit := queryInt(c, "limit", 15, 1, 200)
		order := c.DefaultQuery("order", "abs")
		items, err := service.GetFundingRateTop(c.Request.Context(), d.CH, limit, order)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, gin.H{"items": items})
	}
}

func handleLSVolumeRank(d *Deps) gin.HandlerFunc {
	return func(c *gin.Context) {
		market := queryMarket(c)
		bucket := c.DefaultQuery("bucket", "1h")
		if bucket != "15m" && bucket != "1h" && bucket != "1d" {
			bucket = "1h"
		}
		limit := queryInt(c, "limit", 10, 1, 200)
		result, err := service.GetLongShortVolumeRank(c.Request.Context(), d.Store, d.Cfg, market, bucket, limit)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, result)
	}
}

func handleOIMcapRank(d *Deps) gin.HandlerFunc {
	return func(c *gin.Context) {
		limit := queryInt(c, "limit", 15, 1, 200)
		items, err := service.GetOIMarketCapRank(c.Request.Context(), d.CH, limit)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, gin.H{"items": items})
	}
}
